packages/util/src/emulator.ts

```diff
@@ -140,3 +140,114 @@ export function createMockUserToken(
     signature
   ].join('.');
 }
+
+interface EmulatorStatuses {
+  [name: string]: boolean;
+}
+const emulatorStatus: EmulatorStatuses = {};
+
+interface EmulatorSummary {
+  prod: string[];
+  emulator: string[];
+}
+
+// Checks whether any products are running on an emulator
+function getEmulatorSummary(): EmulatorSummary {
+  const summary: EmulatorSummary = {
+    prod: [],
+    emulator: []
+  };
+  for (const key of Object.keys(emulatorStatus)) {
+    if (emulatorStatus[key]) {
+      summary.emulator.push(key);
+    } else {
+      summary.prod.push(key);
+    }
+  }
+  return summary;
+}
+
+function getOrCreateEl(id: string): { created: boolean; element: HTMLElement } {
+  let parentDiv = document.getElementById(id);
+  let created = false;
+  if (!parentDiv) {
+    parentDiv = document.createElement('div');
+    parentDiv.setAttribute('id', id);
+    created = true;
+  }
+  return { created, element: parentDiv };
+}
+
+/**
+ * Updates Emulator Banner. Primarily used for Firebase Studio
+ * @param name
+ * @param isRunningEmulator
+ * @public
+ */
+export function updateEmulatorBanner(
+  name: string,
+  isRunningEmulator: boolean
+): void {
+  if (
+    typeof window === 'undefined' ||
+    typeof document === 'undefined' ||
+    emulatorStatus[name] === isRunningEmulator
+  ) {
+    return;
+  }
+
+  emulatorStatus[name] = isRunningEmulator;
+  const bannerId = '__firebase__banner';
+  const summary = getEmulatorSummary();
+  const showError = summary.prod.length > 0;
+
+  function tearDown(): void {
+    const element = document.getElementById(bannerId);
+    if (element) {
+      element.remove();
+    }
+  }
+
+  function setupDom(): void {
+    const banner = getOrCreateEl(bannerId);
+    const firebaseText: HTMLSpanElement =
+      document.getElementById('__firebase__text') ||
+      document.createElement('span');
+    if (banner.created) {
+      // update styles
+      const bannerEl = banner.element;
+      bannerEl.style.display = 'flex';
+      bannerEl.style.background = '#7faaf0';
```

Comment:
now that we have multiple of these, can we extract the constants like colors?

Reply:
There's only one now. Do you still want me to name it?

Reply:
Going to resolve this.